blockchain/validate.go

```diff
@@ -1422,55 +1433,109 @@ func checkTreasurybaseUniqueHeight(blockHeight int64, block *dcrutil.Block) erro
+	for _, voteTicketHash := range voteTicketHashes {
+		// Ensure the ticket being spent is actually eligible to vote in this block.
+		if _, ok := winningHashes[voteTicketHash]; !ok {
+			errStr := fmt.Sprintf("block contains vote for "+
+				"ineligible ticket %s (eligible tickets: %s)",
+				voteTicketHash, winners)
+			return ruleError(ErrTicketUnavailable, errStr)
+		}
+
+		// Mark that a vote exists for the ticket hash.
+		winningHashes[voteTicketHash] = true
 	}
 
-	// Determine the winning ticket hashes and create a map for faster lookup.
-	ticketsPerBlock := int(b.chainParams.TicketsPerBlock)
-	winningHashes := make(map[chainhash.Hash]struct{}, ticketsPerBlock)
-	for _, ticketHash := range parentStakeNode.Winners() {
-		winningHashes[ticketHash] = struct{}{}
+	// Create a map of the ticket hashes that will become missed as of the block.
+	missedTicketHashes := make(map[chainhash.Hash]struct{}, len(winningHashes))
+	for ticketHash, hasVote := range winningHashes {
+		// If a winning ticket does not have a vote in this block, it will be
+		// missed.
+		if !hasVote {
+			missedTicketHashes[ticketHash] = struct{}{}
+		}
 	}
+	numMissed := len(missedTicketHashes)
 
-	for _, stx := range block.STransactions {
-		if stake.IsSSGen(stx, isTreasuryEnabled) {
-			// Ensure the ticket being spent is actually eligible to vote in
-			// this block.
-			ticketHash := stx.TxIn[1].PreviousOutPoint.Hash
-			if _, ok := winningHashes[ticketHash]; !ok {
-				errStr := fmt.Sprintf("block contains vote for "+
-					"ineligible ticket %s (eligible tickets: %s)",
-					ticketHash, winningHashes)
-				return ruleError(ErrTicketUnavailable, errStr)
-			}
+	// Create a map of the ticket hashes that will become expired as of the block.
+	numExpiring := len(expiringNextBlock)
+	expiringTicketHashes := make(map[chainhash.Hash]struct{}, numExpiring)
+	for _, ticketHash := range expiringNextBlock {
+		expiringTicketHashes[ticketHash] = struct{}{}
+	}
 
-			continue
+	// Create a map of the ticket hashes that are being revoked in the block.
+	revokedTicketHashes := make(map[chainhash.Hash]struct{},
+		numMissed+numExpiring)
+
+	// Iterate through all revocations in the block, validate that they are
+	// spending tickets that are eligible to be revoked, and add the revocation
+	// to the revocations map.
+	for _, revocationTicketHash := range revocationTicketHashes {
+		// Determine if the ticket being revoked will become missed or expired as of
+		// this block.
+		_, missedInBlock := missedTicketHashes[revocationTicketHash]
+		_, expiringInBlock := expiringTicketHashes[revocationTicketHash]
+		missedOrExpiredInBlock := missedInBlock || expiringInBlock
+
+		// Ensure the ticket being spent is actually eligible to be revoked in this
+		// block.  If the automatic ticket revocations agenda is active, then
+		// revocations are allowed if the ticket will become missed or expired as of
+		// this block.
+		eligible := isAutoRevocationsEnabled && missedOrExpiredInBlock ||
```

Comment:
Yeah, I figured since it's pretty standard across languages that `&&` has precedence over `||` that it read OK, but I don't mind adding parens which makes it immediately obvious.  Updated.